src/kernels/rotary_embedding_kernels.cpp

```diff
@@ -0,0 +1,134 @@
+//     cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+//     sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+//     q_embed = (q * cos) + (rotate_half(q) * sin)
+//     k_embed = (k * cos) + (rotate_half(k) * sin)
+//     return q_embed, k_embed
+//
+
+void llamaApplyRotaryPosEmbed(float *query, float *key, float *emb_cos, float *emb_sin, int qStride, int kStride,
+        int dim, int totSeqLen, int qHeads, int kHeads, const int *positionIds) {
+    const int half = (dim + 1) / 2;
+    const int heads = std::max(qHeads, kHeads);
+
+#pragma omp parallel for collapse(2)
+    for (int head = 0; head < heads; ++head) {
+        for (int seq = 0; seq < totSeqLen; ++seq) {
+            int pos = positionIds[seq];
+            float *pcos = emb_cos + pos * half;
+            float *psin = emb_sin + pos * half;
+
+            float *q = query + seq * qStride + head * dim;
+            float *k = key + seq * kStride + head * dim;
+
+#pragma omp simd
+            for (int i = 0; i < half; ++i) {
+                if (head < qHeads) {
+                    auto q1 = q[i];
+                    q[i] = q1 * pcos[i] - q[i + half] * psin[i];
+                    q[i + half] = q[i + half] * pcos[i] + q1 * psin[i];
+                }
+                if (head < kHeads) {
+                    auto k1 = k[i];
+                    k[i] = k1 * pcos[i] - k[i + half] * psin[i];
+                    k[i + half] = k[i + half] * pcos[i] + k1 * psin[i];
+                }
+            }
+        }
+    }
+}
+
+void llamaApplyRotaryPosEmbed(bfloat16_t *query, bfloat16_t *key, float *emb_cos, float *emb_sin, int qStride,
+        int kStride, int dim, int totSeqLen, int qHeads, int kHeads, const int *positionIds) {
+    const int half = (dim + 1) / 2;
+    const int heads = std::max(qHeads, kHeads);
+
+#pragma omp parallel for collapse(2)
+    for (int head = 0; head < heads; ++head) {
+        for (int seq = 0; seq < totSeqLen; ++seq) {
+            int pos = positionIds[seq];
+            float *pcos = emb_cos + pos * half;
+            float *psin = emb_sin + pos * half;
+
+            bfloat16_t *q = query + seq * qStride + head * dim;
+            bfloat16_t *k = key + seq * kStride + head * dim;
+
+            // Process chunks of 16 elements at a time
+            for (int i = 0; i < half; i += 16) {
+                int remain = half - i;
+                __mmask16 mask = (remain >= 16 ? 0xffff : (1 << remain) - 1);
+
+                __m512 pCosVec = _mm512_maskz_loadu_ps(mask, &pcos[i]);
+                __m512 pSinVec = _mm512_maskz_loadu_ps(mask, &psin[i]);
+
+                // Compute something like:
+                // q[i] = q[i] * pcos[i] - q[i + half] * psin[i];
+                // q[i + half] = q[i + half] * pcos[i + half] + q[i] * psin[i + half];
+                if (head < qHeads) {
+                    __m512 qVec = bfloat16_t::cvt_bf16_to_fp32(_mm256_maskz_loadu_epi16(mask, &q[i]));
```

Comment:
now we can use xft::load_avx512 to make it shorter.

Reply:
Done~